server/routerlicious/packages/services-utils/src/documentKeyRetriever.ts

```diff
@@ -0,0 +1,130 @@
+ */
+
+import { IDocumentManager } from "@fluidframework/server-services-core";
+import { Lumberjack, getLumberBaseProperties } from "@fluidframework/server-services-telemetry";
+
+export class DocumentKeyRetriever {
+	public constructor(
+		// TODO: Consolidate implementations of RedisCache class across all services, then change "any" type to Redis cache
+		private readonly redis: any,
+		private readonly documentManager: IDocumentManager,
+		private readonly loggingEnabled: boolean = true,
+	) {
+		if (!redis) {
+			throw new Error("Redis cache is undefined.");
+		}
+		if (!documentManager) {
+			throw new Error("Document manager is undefined.");
+		}
+	}
+
+	public async getKeyCosmos<ValType>(
+		keyName: string,
+		tenantId: string,
+		documentId: string,
+		useCachedDocument: boolean = true,
+	): Promise<ValType> {
+		this.infoLog(`Retrieving value of ${keyName} from cosmosDB.`, documentId, tenantId);
+
+		// Retrieve the cached document details, if it exists for this document
+		const cachedDetails: Record<string, any> | undefined = useCachedDocument
+			? await this.getCachedDocumentDetails(documentId)
+			: undefined;
+
+		let val: ValType = cachedDetails?.[keyName];
+		if (val) {
+			// If the cached document details contain the key and useCachedDocument is true, use the cached value
+			this.infoLog("Using cached cosmosDB document details.", documentId, tenantId);
+		} else {
+			// Otherwise, call cosmosDB, and cache the result
+			this.infoLog("Querying cosmosDB and caching the results.", documentId, tenantId);
+			const documentDetails: Record<string, any> = await this.documentManager.readDocument(
+				tenantId,
+				documentId,
+			);
+			await this.setCachedDocumentDetails(documentId, documentDetails);
+			val = documentDetails?.[keyName];
+		}
+
+		return val;
+	}
+
+	public async getKeyRedis<ValType>(keyName: string): Promise<ValType> {
+		this.infoLog(`Retrieving value of ${keyName} from redis`);
+		const val: ValType = await this.redis.get(keyName);
+		return val;
+	}
+
+	public async getKeyRedisFallback<ValType>(
+		keyNameBase: string,
+		tenantId: string,
+		documentId: string,
+		useCachedDocument: boolean = true,
+	): Promise<ValType> {
+		const redisKeyName: string = `${keyNameBase}:${documentId}}`;
+		const cosmosKeyName: string = keyNameBase;
+
+		let val: ValType = await this.getKeyRedis<ValType>(redisKeyName);
+		if (val !== undefined && val !== null) {
+			// Return the value if redis has it
+			this.infoLog(`Found value for ${redisKeyName} in redis.`, documentId, tenantId);
+			return val;
+		} else {
+			// Otherwise, get the value from cosmos and cache the result in redis
+			this.infoLog(
+				`Could not find value for ${redisKeyName} in redis. Falling back to cosmosDB for keyname ${cosmosKeyName}.`,
+				documentId,
+				tenantId,
+			);
+
+			val = await this.getKeyCosmos<ValType>(
+				cosmosKeyName,
+				tenantId,
+				documentId,
+				useCachedDocument,
+			);
+			await this.redis.set(redisKeyName, val);
```

Comment:
If I understand correctly, you basically maintained two levels of caching:
1. property key to value
2. document key to value.

In caching, one hard problem is cache inconsistency with db value, and we normally need to build the cache invalidation workflow to make sure cache stay consistence with db

However, here I don't see a cache invalidation, and two levels of caching would make the inconsistency more often which would be a very bad thing. 

Could we simplify the work a bit?